lib/ecto/query/builder.ex

```diff
@@ -815,6 +831,36 @@ defmodule Ecto.Query.Builder do
   def field!(other),
     do: error!("expected atom in field/2, got: `#{inspect other}`")
 
+  defp escape_json_path(path) do
+    Enum.map(path, &quoted_json_path_element!/1)
+  end
+
+  defp quoted_json_path_element!({:^, _, [expr]}),
+    do: quote(do: Ecto.Query.Builder.json_path_element!(unquote(expr)))
+
+  defp quoted_json_path_element!(binary) when is_binary(binary),
+    do: binary
+
+  defp quoted_json_path_element!(integer) when is_integer(integer),
+    do: integer
```

Comment:
Good call on keeping it as an integer here. We probably want to convert this to binary on the adapter.